src/Build/Instance/ProjectItemInstance.cs

```diff
@@ -1791,6 +1791,18 @@ internal void SetMetadataOnTaskOutput(string name, string evaluatedValueEscaped)
                 }
             }
 
+            internal void SetMetadataOnTaskOutput(IEnumerable<KeyValuePair<string, string>> items)
+            {
+                ProjectInstance.VerifyThrowNotImmutable(_isImmutable);
+                _directMetadata ??= new CopyOnWritePropertyDictionary<ProjectMetadataInstance>();
+
+                var metadata = items
+                    .Where(item => !FileUtilities.ItemSpecModifiers.IsDerivableItemSpecModifier(item.Value))
```

Comment:
It should have been `item.Key` instead of `item.Value` - this mistake introduced the problem reported in https://github.com/dotnet/msbuild/issues/8153. I'm working on covering this scenario with unit tests.